crypto/bio/cipher.c

```diff
@@ -0,0 +1,300 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <errno.h>
+#include <openssl/buffer.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/mem.h>
+#include <openssl/pkcs7.h>
+#include <stdio.h>
+#include "../crypto/bio/internal.h"
+#include "../fipsmodule/cipher/internal.h"
+#include "../internal.h"
+#include "./internal.h"
+
+static int enc_write(BIO *h, const char *buf, int num);
+static int enc_read(BIO *h, char *buf, int size);
+static long enc_ctrl(BIO *h, int cmd, long arg1, void *arg2);
+static int enc_new(BIO *h);
+static int enc_free(BIO *data);
+#define ENC_BLOCK_SIZE (1024 * 4)
+
+typedef struct enc_struct {
+  uint8_t done;     // > 0 when finished
+  uint8_t flushed;  // >0 when buffered data has been flushed
+  uint8_t ok;       // decryption status
+  int buf_off;      // start idx of buffered data
+  int buf_len;      // length of buffered data
+  EVP_CIPHER_CTX *cipher;
+  uint8_t buf[ENC_BLOCK_SIZE];
+} BIO_ENC_CTX;
```

Comment:
The block length of the cipher is a value that will need to be considered on each operation. I think it should be in this struct and initialized along with `cipher`.

Reply:
to avoid having to handle this in all possible `EVP_CipherInit` cases, i'd prefer to get the block size with a function call on the cipher. i can do this outside of the loop so it's only called once per invocation of cipher BIO read/write routine.